edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -91,30 +97,110 @@ impl std::fmt::Debug for DockerModuleRuntime {
     }
 }
 
+struct MutexFuture<T>(tokio::sync::lock::Lock<T>);
+
+impl<T> Future for MutexFuture<T> {
+    type Item = tokio::sync::lock::LockGuard<T>;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        Ok(self.0.poll_lock())
+    }
+}
+
 impl ModuleRegistry for DockerModuleRuntime {
     type Error = Error;
     type PullFuture = Box<dyn Future<Item = (), Error = Self::Error> + Send>;
     type RemoveFuture = Box<dyn Future<Item = (), Error = Self::Error>>;
     type Config = DockerConfig;
 
     fn pull(&self, config: &Self::Config) -> Self::PullFuture {
-        let image = config.image().to_string();
-
-        info!("Pulling image {}...", image);
-
-        let creds: Result<String> = config.auth().map_or_else(
-            || Ok("".to_string()),
-            |a| {
-                let json = serde_json::to_string(a).with_context(|_| {
-                    ErrorKind::RegistryOperation(RegistryOperation::PullImage(image.clone()))
-                })?;
-                Ok(base64::encode_config(&json, base64::URL_SAFE))
-            },
-        );
+        let image_with_tag = config.image().to_string();
+        // check if the serveraddress exists & check if it exists in notary_registries
+        let registry_auth = config.auth();
+        let (registry_hostname, registry_username, registry_password) = match registry_auth {
+            Some(a) => (a.serveraddress(), a.username(), a.password()),
+            None => (None, None, None),
+        };
+        let hostname = registry_hostname.unwrap_or_default();
+        let username = registry_username.unwrap_or_default();
+        let password = registry_password.unwrap_or_default();
+
+        let image = if !hostname.is_empty() && !username.is_empty() && !password.is_empty() {
+            let config_path = self.notary_registries.get(hostname);
+            if let Some(c) = config_path {
```

Comment:
You don't need to move c into a temp variable. Just rename c to config_path_buf
```suggestion
            if let Some(config_path_buf) = config_path {
```